network-api/networkapi/wagtailpages/pagemodels/buyersguide/homepage.py

```diff
@@ -523,12 +523,14 @@ def get_product_subset(cutoff_date, authenticated, key, products, language_code=
     if not authenticated:
         products = products.live()
 
-    products = products.prefetch_related(
-        "evaluation__votes",
-        "image__renditions",
-        "product_categories__category",
+    products = (
+        products.prefetch_related(
+            "image__renditions",
+            "product_categories__category",
+        )
+        .annotate(_average_creepiness=Coalesce(models.Avg("evaluation__votes__value"), float(0)))
```

Comment:
**Suggestion:** Perhaps avoid the duplication of this calculation by using a [Prefetch](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#django.db.models.Prefetch) object to retrieve the associated evaluation from an Evaluation queryset using the `with_average_creepiness` method?

Reply:
That's a great suggestion, thanks! I was wondering how I could remove this duplication

Reply:
After trying it, I'm afraid the duplication is inevitable. If we use a `Prefetch` object, it will do another query to prefetch the evaluation objects and then annotate the `_average_creepiness` property onto it. The issue now is that we can't order the original query anymore, i.e.

```python
products = (
        products.prefetch_related(
            "image__renditions",
            "product_categories__category",
            Prefetch("evaluation", ProductPageEvaluation.objects.with_average_creepiness()),
        )
        .order_by("evaluation___average_creepiness")
    )
```
doesn't work since `order_by` won't be able to solve the property in a single query.

Having to reorder the query in Python would be a performance hit, I think, so it justifies leaving the code duplication.

Reply:
Ah got you - the duplication could be warranted here then, although maybe it still makes sense to move it to a method on the Product QuerySet for future convenience?

Reply:
True, a `ProductQuerySet` method would be more convenient and better documented. I'll do that!